### DISCUSSION TITLE
gdscript AST manipulation

### DISCUSSION THREAD

**Discussion by wareya** [godotengine/godot-proposals#4961] [Scripting]
I currently need to do [very, very evil things](https://github.com/wareya/MirageVN/blob/main/singletons/CutsceneParser.gd) to gdscript code on a textual level in my cutscene script loading code, because my cutscene scripts are basically a gdscript-based DSL. This works, but because it works on pure text, it's nasty. If I had a way to parse, manipulate, and compile ASTs instead of pure text, even if it was dictly-typed, it would be a *lot* less fragile.

Metaprogramming at this level is applicable across the board, not just to cutscenes:

- being able to procedurally build or alter AI code would make it easier to do evil un-state-machine-like things without creating tons of spaghetti
- shmup bullet patterns: procedurally inserting timing and pattern stage checks, instead of hardcoding it, makes them less fragile
- games with user generated content that might contain code would have an easier time making sure UGC code sticks to only whitelisted features (this is the wrong place to do security checks, but it would be helpful as an additional defense)
- there are a number of of high-level functional programming techniques that gdscript currently can't do, but that would be possible to *emulate* if AST manipulation were possible
- developers could run custom preprocessors that work around annoyances they might have with gdscript (e.g. adding a substitute for python slice notation by reusing rare syntax that currently has a different meaning (e.g. turning `array[var1--var2]` into `array.slice(var1, var2)`))
- it would be easier to autogenerate inner class types used for data storage from schema data than it currently is
- people could write primitive linting tools entirely within their own project, without having to run any non-godot software, not even as an optional dependency to an addon (e.g. if your team for some reason has a style rule that forbids directly accessing autoloads/singletons from inner classes, you could do that, without having to teach a real linter how to detect it)

**Comment by theraot** [comment]
Perhaps a public wrapper on `GDScriptTokenizer` would do?

**Comment by displague** [comment]
Connecting some dots on search results:
- https://github.com/godotengine/godot-proposals/issues/4958
- https://store.godotengine.org/asset/star-weaver/gdscript-ast/
- https://bitbucket.org/arlez80/godot-byte-code-parser/ (https://www.reddit.com/r/godot/comments/tj3vx7/gdscript_bytecodeast_parser/)
- https://github.com/can1357/oh-my-pi/issues/3366

**Reply by tomsterBG** [reply]
Also this sounds a lot like what Nim can do, for the curious of what is possible:
- https://nim-lang.org/docs/macros.html#dumpTree.m%2Cuntyped - Just dumps an AST
- https://nim-lang.org/docs/macros.html#dumpAstGen.m%2Cuntyped - dumps code to generate given AST
- https://nim-lang.org/docs/tut3.html#introduction-with-power-comes-responsibility - a very important note on macros

**Comment by HolonProduction** [comment]
Similar stuff has been discussed in the context of custom annotations. There are currently no plans to expose the AST in a stable way, let alone allowing mutations. So I'm going to close this.
